[bgp-gr-helper] Add script for testing the BGP graceful restart helper function

wangxin opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Summary:
Fixes # (issue)
Add script for testing the BGP graceful restart helper function.

### Type of change

- [] Bug fix
- [] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Approach
#### How did you do it?
Two scenarios are covered:
* Kill ribd, bring down/up interface on VM to simulate graceful restart. Verify that the DUT preserves the routes.
* Reboot the VM to perform a normal restart. Verify that the DUT does not preserve the routes.

#### How did you verify/test it?
Tested on Mellanox platform

#### Any platform specific information?
N/A

#### Supported testbed topology if it's a new test case?
[t1, t1-lag, t1-64-lag]

### Documentation 
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->
[bgp-gr-helper.log](https://github.com/Azure/sonic-mgmt/files/3365109/bgp-gr-helper.log)

pavel-shirshov commented on ansible/roles/test/tasks/bgp_gr_helper.yml:
```diff
@@ -0,0 +1,127 @@
+- fail: msg="testbed_type is not defined."
+  when: testbed_type is not defined
+
+- fail: msg="testbed_type {{testbed_type}} is invalid."
```
"unsupported"?

wangxin replied:
Sure, corrected in a new commit.

pavel-shirshov commented on ansible/roles/test/tasks/bgp_gr_helper.yml:
```diff
@@ -0,0 +1,127 @@
+    - name: Verify bgp session is established
+      assert: { that: "'{{ bgp_neighbors[peer_ip]['state'] }}' == 'established'" }
+
+    - name: Verify that IPv4 unicast routes were preserved during GR.
```
should we check ipv6 addresses too?

wangxin replied:
Good suggestion. I'll add ipv6 checking.

pavel-shirshov commented on ansible/roles/test/tasks/bgp_gr_helper.yml:
```diff
@@ -0,0 +1,127 @@
+    - name: Verify bgp session is established
+      assert: { that: "'{{ bgp_neighbors[peer_ip]['state'] }}' == 'established'" }
+
+    - name: Verify that IPv4 unicast routes were not preserved during GR. FIB should be updated.
```
Why the routes weren't preserved in this case?

wangxin replied:
In this case, the VM was rebooted. It was not a graceful restart and routes were not supposed to be preserved.

pavel-shirshov replied:
Can you add some comment message to make this case clear?
Or is it possible to make two different cases to easily distinct two cases from each other?

wangxin replied:
Sure, I'll add more comments and make the script looks like two obvious cases.

pavel-shirshov commented on ansible/roles/eos/templates/t1-64-lag-tor.j2:
```diff
@@ -100,6 +100,9 @@ interface {{ bp_ifname }}
 router bgp {{ host['bgp']['asn'] }}
  router-id {{ host['interfaces']['Loopback0']['ipv4'] |  ipaddr('address') }}
  !
+ graceful-restart restart-time {{ bgp_gr_timer }}
```
where is this timer is defined?

wangxin replied:
Good catch, thanks!
This bgp_gr_timer is used when deploy VMs for testing. It is defined in sonic-mgmt/ansible/group_vars/eos/eos.yml. I missed this file in the PR. I will add it in new commits.

pavel-shirshov commented on ansible/roles/test/tasks/bgp_gr_helper/routes_update_expect.txt:
```diff
@@ -0,0 +1 @@
+r, ".*Create route.*"
```
Can you please describe what does this rule describe?

wangxin replied:
When normal reboot is performed on VM, the peer device (SONiC) is not supposed to preserve the routes. After the VM is up, SONiC should learn BGP routes from VM again. This kind of logs are expected in /var/log/syslog:

Jul 18 07:21:24.264802 mtbc-sonic-01-2410 INFO swss#orchagent: :- addRoute: Create route 192.168.14.0/32 with next hop(s) 10.0.0.1,10.0.0.5,10.0.0.9,10.0.0.13,10.0.0.21,10.0.0.25,10.0.0.29
Jul 18 07:21:24.265080 mtbc-sonic-01-2410 INFO swss#orchagent: :- addRoute: Create route 192.168.14.1/32 with next hop(s) 10.0.0.1,10.0.0.5,10.0.0.9,10.0.0.13,10.0.0.21,10.0.0.25,10.0.0.29
Jul 18 07:21:24.265364 mtbc-sonic-01-2410 INFO swss#orchagent: :- addRoute: Create route 192.168.14.112/32 with next hop(s) 10.0.0.1,10.0.0.5,10.0.0.9,10.0.0.13,10.0.0.21,10.0.0.25,10.0.0.29

pavel-shirshov replied:
So here you expect to see only "Create route" messages.

wangxin replied:
Yes, the exact logic here is that: if pattern ".*Create route.*" is not observed in syslog, fail the case.

pavel-shirshov commented on ansible/roles/test/tasks/bgp_gr_helper/routes_update_match.txt:
```diff
@@ -0,0 +1,3 @@
+r, ".*Remove route.*"
+r, ".*Create route.*"
+r, ".*Set route.*"
```
Can you please describe what does this rule describe?

wangxin replied:
When graceful restart is performed on VM, the peer device (SONiC) should preserve the routes. None of the log related with removing or adding route should be observed in /var/log/syslog.

pavel-shirshov replied:
But here (if I understand right) you expect (following logic from the previous comment) to see "Remove route", "Create route", and "Set route" messages.
Did I miss something?

wangxin replied:
The exact logic is that: If any of the below patterns is observed in /var/log/syslog while testing graceful restart, fail the test:
* `.*Remove route.*`
* `.*Create route.*`
* `.*Set route.*`

wangxin commented:
@pavel-shirshov Added new comments addressing the review comments:

* Add IPv6 route check
* Add bgp_gr_timer definition which was missed in previous commits
* Add graceful restart support in t1-tor.j2, which was missed in previous commits

Script execution log is attached too.

[bgp-gr-helper2.log](https://github.com/Azure/sonic-mgmt/files/3405801/bgp-gr-helper2.log)

wangxin commented:
@pavel-shirshov Review comments addressed in the new commit. Could you please check? Thanks!

rajneeshaec commented:
Hi All,

   I am getting following error given below while running the bgp_gr_helper test suite:-

   TASK [test : Get VM GR timer.] *************************************************
task path: /var/azure/sonic-mgmt/ansible/roles/test/tasks/bgp_gr_helper.yml:17
Wednesday 24 July 2019  10:00:09 +0000 (0:00:02.445)       0:01:50.079 ********
fatal: [et6448m-M0]: FAILED! => {"failed": true, "msg": "ERROR! 'dict object' has no attribute 'capabilities'"}

PLAY RECAP *********************************************************************

How to fix this issue?

Thanks
Rajneesh
